common/src/main/java/io/pravega/common/io/serialization/VersionedSerializer.java

```diff
@@ -0,0 +1,654 @@
+    /**
+     * The Version of the Serializer format itself.
+     * ALL               : Header|Version(1)|RevisionCount(1)|Revision1|...|Revision[RevisionCount]
+     * HEADER(SingleType): SFV(1)
+     * HEADER(MultiType) : SFV(1)|ObjectType(1)
+     * REVISION          : RevisionId(1)|Length(4)|Data(Length)
+     */
+    private static final int SERIALIZER_VERSION = 0;
+
+    /**
+     * Serializes the given object to the given RevisionDataOutput. This overload is usually invoked for serializing
+     * nested classes or collections.
+     *
+     * @param dataOutput The RevisionDataOutput to serialize to.
+     * @param object     The object to serialize.
+     * @throws IOException If an IO Exception occurred.
+     */
+    public void serialize(RevisionDataOutput dataOutput, T object) throws IOException {
+        serialize(dataOutput.getBaseStream(), object);
+    }
+
+    /**
+     * Serializes the given object to an in-memory buffer (RandomOutput) and returns a view of it.
+     *
+     * @param object The object to serialize.
+     * @return An ArrayView which represents the serialized data. This provides a view (offset+length) into a Java byte
+     * array and has APIs to extract or copy the data out of there.
+     * @throws IOException If an IO Exception occurred.
+     */
+    public ByteArraySegment serialize(T object) throws IOException {
+        val result = new EnhancedByteArrayOutputStream();
+        serialize(result, object);
+        return result.getData();
+    }
+
+    /**
+     * Serializes the given object to the given OutputStream.
+     *
+     * @param stream The OutputStream to serialize to.
+     * @param object The object to serialize.
+     * @throws IOException If an IO Exception occurred.
+     */
+    public void serialize(OutputStream stream, T object) throws IOException {
+        stream.write(SERIALIZER_VERSION);
+        serializeContents(stream, object);
+    }
+
+    /**
+     * When implemented in a derived class, this method will write the serialization contents, excluding the Header.
+     * Refer to the format above for contents.
+     * @param baseStream The OutputStream to write to.
+     * @param object     The object to serialize.
+     * @throws IOException If an IO Exception occurred.
+     */
+    abstract void serializeContents(OutputStream baseStream, T object) throws IOException;
+
+    /**
+     * Reads a single unsigned byte from the given InputStream and interprets it as a Serializer Format Version, after
+     * which it validates that it is supported.
+     *
+     * @param dataInput The InputStream to read from.
+     * @throws IOException If an IO Exception occurred.
+     */
+    void processHeader(InputStream dataInput) throws IOException {
+        int formatVersion = dataInput.read();
+        if (formatVersion < 0) {
+            throw new EOFException();
+        }
+        ensureCondition(formatVersion == SERIALIZER_VERSION, "Unsupported format version %d.", formatVersion);
```

Comment:
This is the version of the serializer format itself. Nobody will touch this. When we decide to change this, we'll need implement the new format so that the code will know how to read it, and then switch to it. 

Just to clarify: this is not the user-supplied version; this is so that we can change the version of the serialization format itself (see `SERIALIZER_VERSION` and its Javadoc above).